File: kolakoski.c
/******************************************************************************
Calculate the Kolakoski Sequence

by James McClanahan, W4JBM

Original Version in BASIC written in May 2018.

C version written in February 2019.

*******************************************************************************/

#include <stdio.h>

int main ()
{

/*  i array holds the sequence; first two elements are initilized
    p1 points into the array where we are now "getting" values
    p2 points into the array where we are now "putting" values
    n1 is the current "value" loading into the array
    n2 is the other "value" and we "swap" them as needed
    x is used as a temporary index in the for() loop
*/

  int i[60] = { 0, 1, 2 };
  int n1 = 2, n2 = 1, p1 = 2, p2 = 2, x;

  while (p2 < 57)
    {
      /* We start by expanding i[] with p2 as the pointer to the last
         known value and p1 as the pointer to the last value processed
         in the run length encoding. */

      for (x = 1; x <= i[p1]; x++)
	{
	  i[p2 + x - 1] = n1;
	}

      /* After the new values have been "decoded" from the run length
         encoding, p2 is updated to the new end of the decoded portion
         of the array and p1 is updated to point to the new 'last
         location decoded' in the array. */

      p2 = p2 + i[p1];
      p1++;

      /* n1 is the value being loaded into the array and n2 is The
         'other' value. We swap them before we go through the next round
         of decoding. */

      swap (&n1, &n2);

      /* The while() loop continues as long as we have some array
         space remaining. Since we don't know if the value last decoded
         will be a 1 or 2, the array is just oversized to avoid having
         to implement any additional truncation logic. */

    }

  /* Once we have reached "the end", print out the values in the
     array. */

  printf ("The Kolakoski sequence decoded to %d places is:\n", p2 - 1);
  for (x = 1; x < p2; x++)
    {
      printf ("%d", i[x]);
    }

  printf ("\n");


  return 0;
}


/* SWAP function is called by 'swap(&x , &y);' */

void
swap (int *a, int *b)
{
  int t;

  t = *b;
  *b = *a;
  *a = t;
}

File: kolakoski.py
"""
Calculate Kolakoski Sequence in Python

by James McClanahan, W4JBM

Original Version in BASIC written in May 2018.

Python version written in February 2019.

"""

#initialize list with K sequence
#(easier to start at element 1 and waste element 0)

i =[0, 1, 2, 2]

# with Python, it turned out to be easier to initialize
# the first three elements of the sequence because while
# use an append() function to add to a list instead of
# using absolute indexing to place values in an array

n1 = 1
n2 = 2
p1 = 3

## my sense of symmetry likes that last set of constants...

while len(i) < 55:
    for x in range(0, i[p1]):
        i.append(n1)       #add new decoded items to list

    ## for debugging or to watch the list build, uncomment
    ## the following print()
    ## print(i[1:len(i)])

    p1 = p1 +1
    n1, n2 = n2, n1         #swap n1 and n2
    
print("Kolakkoski Sequence to %d digits:" % len(i))
for x in range (1, len(i)):
  print (i[x], end="")
print("")

